src/main/scala/io/iohk/ethereum/jsonrpc/EthProofService.scala

```diff
@@ -0,0 +1,158 @@
+import io.iohk.ethereum.domain.{Account, Address, Block, Blockchain, UInt256}
+import io.iohk.ethereum.jsonrpc.EthService._
+import io.iohk.ethereum.mpt.{MptNode, MptTraversals}
+import monix.eval.Task
+
+/** The key used to get the storage slot in its account tree */
+final case class StorageProofKey(v: BigInt) extends AnyVal
+
+/**
+  * Object proving a relationship of a storage value to an account's storageHash
+  *
+  * @param key storage proof key
+  * @param value the value of the storage slot in its account tree
+  * @param proof the set of node values needed to traverse a patricia merkle tree (from root to leaf) to retrieve a value
+  */
+final case class StorageProof(
+    key: StorageProofKey,
+    value: BigInt,
+    proof: Seq[ByteString]
+)
+
+/**
+  * The merkle proofs of the specified account connecting them to the blockhash of the block specified.
+  *
+  * Proof of account consists of:
+  * - account object: nonce, balance, storageHash, codeHash
+  * - Markle Proof for the account starting with stateRoot from specified block
+  * - Markle Proof for each requested storage entory starting with a storage Hash from the account
+  *
+  * @param address the address of the account or contract of the request
+  * @param accountProof Markle Proof for the account starting with stateRoot from specified block
+  * @param balance the Ether balance of the account or contract of the request
+  * @param codeHash the code hash of the contract of the request (keccak(NULL) if external account)
+  * @param nonce the transaction count of the account or contract of the request
+  * @param storageHash the storage hash of the contract of the request (keccak(rlp(NULL)) if external account)
+  * @param storageProof current block header PoW hash
+  */
+final case class ProofAccount(
+    address: Address,
+    accountProof: Seq[ByteString],
+    balance: BigInt,
+    codeHash: ByteString,
+    nonce: UInt256,
+    storageHash: ByteString,
+    storageProof: Seq[StorageProof]
+)
+
+sealed trait MptProofError
+case object UnableRebuildMpt extends MptProofError
+case object KeyNotFoundInRebuidMpt extends MptProofError
+
+/**
+  * Spec: [EIP-1186](https://eips.ethereum.org/EIPS/eip-1186)
+  * besu: https://github.com/PegaSysEng/pantheon/pull/1824/files
+  * parity: https://github.com/openethereum/parity-ethereum/pull/9001
+  * geth: https://github.com/ethereum/go-ethereum/pull/17737
+  */
+class EthProofService(blockchain: Blockchain, blockGenerator: BlockGenerator, ethCompatibleStorage: Boolean) {
+
+  /**
+    * Get account and storage values for account including Merkle Proof.
+    *
+    * @param address address of the account
+    * @param storageKeys storage keys which should be proofed and included
+    * @param blockNumber block number or string "latest", "earliest"
+    * @return
+    */
+  def run(
+      address: Address,
+      storageKeys: Seq[StorageProofKey],
+      blockNumber: BlockParam
```

Comment:
would it be possible to name this param in a less clashing way?

Reply:
Renamed `blockNumber` to `block`. WDYT ?

PS JSON-RPC spec uses `BlockNumber`: https://github.com/etclabscore/ethereum-json-rpc-specification/blob/master/openrpc.json#L717 as a name of type. We have BlockParam (that is more accurate as this can be number or description like latest).